---
title: "HRECs-AMD-Cytokine-RNASeq"
author: "Fergus McLellan"
date: "`r Sys.Date()`"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Setup Space and Load Data
  
 clear space

```{r settings}
rm(list = ls(all.names = TRUE))
gc()
options(max.print = .Machine$integer.max, scipen = 999, stringsAsFactors = F, dplyr.summarise.inform = F)
```



setwd("C:/Users/User/OneDrive - UNSW/Desktop/Honours/")

set.seed(12345)

Load initial packages and paths

```{r paths}
library(tidyverse) #tidyverse used for code cleanliness and general functions

in_path <- "C:/Users/User/OneDrive - UNSW/Desktop/Honours/raw_data/" #set input path
out_path = "C:/Users/User/OneDrive - UNSW/Desktop/Honours/fibrosis_figure_panel/significant_data/" # set output path
```

# DESeq2 ----------------------------------------------------


Differential expression analysis using DESEQ2

There are 3 objects required for DESeq to run properly.

1.A Counts matrix, with one 1 set of labels (Ensembl ID in this case) as the row name

2. A Design factor with 1 level per treatment, and repetitions = n

3. A colData data frame containing a column reflecting the design


 1. Counts matrix

```{r}
Counts <-read.delim(paste0(in_path,"raw_counts.csv"), header = TRUE, row.names = 1, sep = ",")

Counts <- as.matrix(Counts[,1:48])

head(Counts)
```

  2 & 3
    In this case, I have generated the function which allows for a few preset analyses of interest so as to keep the       global environment neat.

```{r}

#Possible comparison types: all_groups, fibro, inflam, angio, single *IF SINGLE MUST ALSO IDENTIFY TREATMENT*

metadatapicker <- function(comparisontype, treatment = NULL){

  
  
  if(comparisontype == "important"){
    condition <- factor(c(rep("Control", 6), rep("TGFb2", 6), rep("TNFa", 6), rep("Thrombin", 6), rep("ALL", 6)))
    coldata <- data.frame(condition)
    coldata$condition <- relevel(coldata$condition, ref = "Control")
    
    # All Condition ---------------------------------------------------------
  
    }else if (comparisontype == "all"){
    
    condition <- factor(c(rep("Control", 6), rep("TGFb1", 6), rep("TGFb2", 6), rep("TNFa", 6), rep("Thrombin", 6) , rep("IL6", 6), rep("VEGF", 6), rep("ALL", 6)))
    coldata <- data.frame(condition)
    coldata$condition <- relevel(coldata$condition, ref = "Control")    
    
    # Fibro Condition ---------------------------------------------------------
  
    }else if (comparisontype == "fibro"){
    
    condition <- factor(c(rep("Control", 6), rep("TGFb1", 6), rep("TGFb2", 6), rep("TNFa", 6), rep("Thrombin", 6)))
    coldata <- data.frame(condition)
    coldata$condition <- relevel(coldata$condition, ref = "Control")   
    
    # Inflam Condition --------------------------------------------------------
  
    }else if (comparisontype == "inflam"){
    
    condition <- factor(c(rep("Control", 6), rep("TGFb1", 6), rep("TNFa", 6), rep("Thrombin", 6) , rep("IL6", 6)))
    coldata <- data.frame(condition)
    coldata$condition <- relevel(coldata$condition, ref = "Control")    
    
    # Angio Condition ---------------------------------------------------------
  
    }else if (comparisontype == "angio"){
    
    condition <- factor(c(rep("Control", 6), rep("TGFb2", 6), rep("Thrombin", 6) ,rep("VEGF", 6)))
    coldata <- data.frame(condition)
    coldata$condition <- relevel(coldata$condition, ref = "Control")    
    
    # Individuals only (no cocktail) ---------------------------------------------------------
    
    }else if (comparisontype == "individuals"){
    condition <- factor(c(rep("Control", 6), rep("TGFb1", 6), rep("TGFb2", 6), rep("TNFa", 6), rep("Thrombin", 6) , rep("IL6", 6), rep("VEGF", 6)))
    coldata <- data.frame(condition)
    coldata$condition <- relevel(coldata$condition, ref = "Control")
    
    
    # Single Comparison Condition ---------------------------------------------
  
    }else if (comparisontype == "single"){
    
      if(!is.null(treatment)){
      condition <- factor(c(rep("Control", 6), rep(treatment, 6)))
      coldata <- data.frame(condition)
      coldata$condition <- relevel(coldata$condition, ref = "Control")
    
      }else{
      print("Please select a treatment for your single comparison. Returning only Control group in data frame")
        }
  
      }else{
    print("Invalid selection. Please choose from the list of possible comparisons.")
    return(NULL)
  }
  
  return(list("condition" = condition, "coldata" = coldata))

}
```

Use function to set meta data 
allows: all, individuals, fibro, inflam, angio and single* 
  *single comparisons requires selection of the treatment of interest


assign a list for the meta data objects and split this list to give a condition factor and coldata dataframe object
```{r}
metadata <- metadatapicker(comparisontype = "all",
                         treatment =
                          ) %>%
               list2env(.GlobalEnv)
```

2. Condition factor

```{r}
levels(condition)
```
3. colData data frame

```{r}
head(coldata)
```


Next, it is best practice to clean the data to remove any NA values, or unusable counts.
For this I have used the WGCNA package goodSamplesGenes function.

```{r}

ddsCounts <- data

library(WGCNA)


gsg <- goodSamplesGenes(t(ddsCounts))
summary(gsg)

#false in all ok is bad
gsg$allOK
#is it genes that are bad (iterative check for repeated 0 values)?
table(gsg$goodGenes)
```

#is it samples that are bad?
```{r}
table(gsg$goodSamples)
```
Clean ddsCounts based on gsg findings
```{r}
ddsCounts <- ddsCounts[gsg$goodGenes == TRUE,]
```

Run DESeq2
```{r}

library(DESeq2)

dds <- DESeqDataSetFromMatrix(countData = ddsCounts, colData = coldata, design= ~important_groups_condition)

dds <- DESeq(dds)

```


```{r}
res_norm <- as.data.frame(counts(dds, normalized = TRUE))
nrow(res_norm2)
rowSums(res_norm2)


#res_norm2Clean <- res_norm2[rowSums(res_norm2 > 6)>=6,]
#nrow(res_norm2Clean)

write.csv(res_norm, file = paste0(out_path,"important_groups_ggs_nocutoff_normalised_001.csv"))
```

```{r}

```




















out_path <- "fibrosis_figure_panel/volcano_plots/"
